cinder/volume/drivers/nexenta/ns5/iscsi.py

```diff
@@ -133,38 +136,156 @@ def __init__(self, *args, **kwargs):
         except jsonrpc.NefException as error:
             if error.code != 'ENOENT':
                 raise
-            block_size = self.configuration.nexenta_ns5_blocksize
-            if block_size < 512:
-                block_size *= units.Ki
-            payload = {
-                'path': self.root_path,
+            block_size = self.configuration.nexenta_blocksize
+            create_payload = {
+                'path': self.san_path,
                 'volumeBlockSize': block_size
             }
-            self.nef.volumegroups.create(payload)
+            self.nef.volumegroups.create(create_payload)
+            self.san_stat = self.nef.volumegroups.get(self.san_path,
+                                                      stat_payload)
         service = self.nef.services.get('iscsit')
         if service['state'] != 'online':
             message = (_('iSCSI target service is not online: %(state)s')
                        % {'state': service['state']})
             raise jsonrpc.NefException(code='ESRCH', message=message)
 
+    def _get_volume_reservation(self, volume):
+        """Calculates the correct reservation size for given volume size.
+
+        Its purpose is to reserve additional space for volume metadata
+        so volume don't unexpectedly run out of room. This function is
+        a copy of the volsize_to_reservation function in libzfs_dataset.c
+
+        :param volume: volume reference
+        :returns: reservation size
+        """
+        volume_path = self._get_volume_path(volume)
+        payload = {'fields': 'volumeBlockSize,volumeSize,dataCopies'}
+        volume_specs = self.nef.volumes.get(volume_path, payload)
+        block_size = volume_specs['volumeBlockSize']
+        volume_size = volume_specs['volumeSize']
+        data_copies = volume_specs['dataCopies']
+        reservation = volume_size
+        numdb = 7
+        dn_max_indblkshift = 17
+        spa_blkptrshift = 7
+        spa_dvas_per_bp = 3
+        dnodes_per_level_shift = dn_max_indblkshift - spa_blkptrshift
+        dnodes_per_level = 1 << dnodes_per_level_shift
+        nblocks = reservation // block_size
+        while nblocks > 1:
+            nblocks += dnodes_per_level - 1
+            nblocks //= dnodes_per_level
+            numdb += nblocks
+        numdb *= min(spa_dvas_per_bp, data_copies + 1)
+        reservation *= data_copies
+        numdb *= 1 << dn_max_indblkshift
+        reservation += numdb
+        volume_meta = reservation - volume_size
+        LOG.debug('Reservation size for raw volume %(volume)s: '
+                  '%(reservation)s, volume data size: %(volume_size)s '
+                  'and volume metadata size: %(volume_meta)s',
+                  {'volume': volume['name'], 'reservation': reservation,
+                   'volume_size': volume_size, 'volume_meta': volume_meta})
+        return reservation
+
+    def _set_volume_reservation(self, volume, reservation=None):
```

Comment:
Could you please add a comment to explain why do we need to set volsize several times during this operation?

Reply:
Sure, please let me add a comment. Thank you for the review!

Reply:
done